doc/testnet_configuration.md
# Epic Server - Floonet Configuration

## Objective

This document aims to show you how to become a contributor in the testing area and help Epic to be an increasingly great cryptocurrency.

## Requirements

Before starting the settings, we first need to have installed the most current version of Epic's code on GitHub. 

To do this, just follow what is described in [README.md](https://github.com/EpicCash/epic/#readme) or in [build.md](https://github.com/EpicCash/epic/blob/master/doc/build.md).

Once you have all Node, Wallet and Miner services up and running, just follow this document to configure your environment to connect to the test environment.

## How to Access the Testnet

After having to download the latest version of the Epic Server, Wallet, and Miner. You need following the steps:

### Reset the Server Configuration

If the testnet is restarted or there's a new version of the epic server, you will need to remove the directory called **/chain/data**. This directory is where the epic cash blockchain stores its data. 

Therefore, if the testnet is restarted, all this data needs to be removed in order to run and store the newest version of the blockchain. The following steps explain how to erase this data using the terminal:

1. Open a new terminal window in the directory where you saved the epic server data. If you used the [default configuration](./running.org#epic_config_default), this folder should be under **~/.epic/floo** in you home directory.
2. Then execute the following command:

   ```sh
   rm -rf chain_data/
   ```
After that check, if you have the following settings for your node to connect to the other floonet servers:

#### Windows

Download all binaries from the requirements list above and extract them in a convenient location (e.g C:\Program Files\Epic).

1. Go to the directory to store the configuration files:

  Using PowerShell:

  ```shell
  cd $HOME\.epic\floo
  ```

  Using CMD:

  ```cmd
  cd $USERPROFILE%\.epic\floo
  ```

2. Open the `epic-server.toml` file and change the value of `seeding_type` from `DNSSeed` to `List`:

  ```toml
  #how to seed this server, can be None, List or DNSSeed
  seeding_type = "List"
  ```

4. And add the following peers to the seeds list:

  ```toml
  #If the seeding type is List, the list of peers to connect to can
  #be specified as follows:
  seeds = ["15.229.31.27:23414"]
  ```
  Save the `epic-server.toml` file.

5. Start the server and wait until it is fully synchronized with the other peers:

  ```shell
  epic.exe --floonet
  ```

#### Linux

These are the steps to reset your `floonet`:

1. Go to the directory to store the configuration files:

  ```bash
  cd ~/.epic/floo
  ```

2. Open the `epic-server.toml` file and change the value of `seeding_type` from `DNSSeed` to `List`:

  ```toml
  #how to seed this server, can be None, List or DNSSeed
  seeding_type = "List"
  ```

3. And add the following peers to the seeds list:

  ```toml
  #If the seeding type is List, the list of peers to connect to can
  #be specified as follows:
  seeds = ["15.229.31.27:23414"]
  ```
  Save the `epic-server.toml` file.

4. Start the server and wait until it is fully synchronized with the other peers:

  ```bash
  epic --floonet
  ```

### First Run of Epic Server

#### Windows

Download all binaries from the requirements list above and extract them in a convenient location (e.g C:\Program Files\Epic).

1. Create the directory to store the configuration files:

  Using PowerShell:

  ```shell
  md -f $HOME\.epic\floo

  cd $HOME\.epic\floo
  ```

  Using CMD:

  ```cmd
  md $USERPROFILE%\.epic\floo

  cd $USERPROFILE%\.epic\floo
  ```

2. Create the server configuration files:

  ```shell
  epic.exe --floonet server config
  ```

3. Open the `epic-server.toml` file and change the value of `seeding_type` from `DNSSeed` to `List`:

  ```toml
  #how to seed this server, can be None, List or DNSSeed
  seeding_type = "List"
  ```

4. And add the following peers to the seeds list:

  ```toml
  #If the seeding type is List, the list of peers to connect to can
  #be specified as follows:
  seeds = ["15.229.31.27:23414"]
  ```
  Save the `epic-server.toml` file.

5. Create the flag file:

  Using PowerShell:

  ```shell
  New-Item 2-15-rollback-flag
  ```

  Using cmd:

  ```shell
  type nul > 2-15-rollback-flag
  ```

6. Start the server and wait until it is fully synchronized with the other peers:

  ```shell
  epic.exe --floonet
  ```

#### Linux

These are the steps to access the `testnet`:

1. Create the directory to store the configuration files:

  ```bash
  mkdir -p ~/.epic/floo && cd ~/.epic/floo
  ```

2. Create the server configuration file:

  ```bash
  epic --floonet server config
  ```

3. Open the `epic-server.toml` file and change the value of `seeding_type` from `DNSSeed` to `List`:

  ```toml
  #how to seed this server, can be None, List or DNSSeed
  seeding_type = "List"
  ```

4. And add the following peers to the seeds list:

  ```toml
  #If the seeding type is List, the list of peers to connect to can
  #be specified as follows:
  seeds = ["15.229.31.27:23414"]
  ```
  Save the `epic-server.toml` file.

5. Create the flag file:

  ```bash
  touch 2-15-rollback-flag
  ```

6. Start the server and wait until it is fully synchronized with the other peers:

  ```bash
  epic --floonet
  ```

## How to Start Mining on Testnet

### Windows

1. Still on the same `.epic/floo` directory, initiate the wallet:

  ```shell
  epic-wallet.exe --floonet init -h
  ```

  You'll be prompted to enter the password for the wallet.

2. Create a default account:

  ```shell
  epic-wallet.exe --floonet account
  ```

3. Double check the initial balance:

  ```shell
  epic-wallet.exe --floonet info
  ```

4. Start listening for connections on the `epic-wallet`:

  ```shell
  epic-wallet.exe --floonet listen
  ```

5. Configure your miner to work with the `floonet` server by changing the port of the stratum server on the `epic-miner.toml` file.
   This file is located in the folder where you extracted the `epic-miner` files.

  ```toml
  # listening epic stratum server url
  stratum_server_addr = "127.0.0.1:13416"
  ```

6. Open a new terminal, then go to the folder where you extracted the `epic-miner` and execute it:

  ```shell
  epic-miner.exe
  ```

  or

  ```shell
  epic-miner-opencl.exe
  ```

### Linux

1. Still on the same `.epic/floo` directory, initiate the wallet:

  ```bash
  epic-wallet --floonet init -h
  ```

  You'll be prompted to enter the password for the wallet.

2. Create a default account:

  ```bash
  epic-wallet --floonet account
  ```

3. Double check the initial balance:

  ```bash
  epic-wallet --floonet info
  ```

4. Start listening for connections on the `epic-wallet`:

  ```bash
  epic-wallet --floonet listen
  ```

5. Configure your miner to work with the `floonet` server by changing the port of the stratum server on the `/opt/epic-miner/epic-miner.toml` file:

  ```toml
  # listening epic stratum server url
  stratum_server_addr = "127.0.0.1:13416"
  ```

  The example above is related to the CPU-only miner.

6. Start the miner:

  ```bash
  epic-miner
  ```

CHANGELOG.md
# epic 3.0.9-alpha-1

  - Add --onlyrandomx and --noprogpow for testnet
  - Fix: mining on usernet
  - Feat: speed up mine blocks on usernet
  - Fix: Variable Reference
  - Update build.md
  - Create testnet_configuration.md
  - Feat: Add get_blocks and get_last_n_kernels
  - [#21](https://github.com/EpicCash/epic/issues/21) Add simple print to indicate the foundation is broke
  - Fix: cucumber tests
  - Add CDN


# epic 3.0.0

  - [#50](https://github.com/EpicCash/epic/issues/50) Fix: Address issue Breaks windows build
  - [#46](https://github.com/EpicCash/epic/issues/46) Merge pull request  from acosta-brickabode/feature/gh-action-linux
  - [#45](https://github.com/EpicCash/epic/issues/45) Merge pull request  from acosta-brickabode/fix/log-level-syncer
  - Fix: Replace pancurses-backend feature with termion-backend to fix ncurses rendering issues
  - Refactor: Remove libssl-dev as build requirement
  - Fix: Replace log::info feature with log::debug to fix overprinting sync messages
  - Feat: Implement Github Action for generating Linux artifact
  - [#40](https://github.com/EpicCash/epic/issues/40) Doubling the value of the MAINNET_FIRST_HARD_FORK constant 
  - Fix: Update croaring to a fork that doesn't use `-march=native` that was causing the illegal instruction errors
  - [#38](https://github.com/EpicCash/epic/issues/38) Retry connecting to the seed servers when there is no connected peers
  - Test(config): Add unit tests to check compatibility with v2 and v3 values for log-levels
  - Fix(config): Use custom deserialization function that accepts log-level values from both v2 and v3
  - [#35](https://github.com/EpicCash/epic/issues/35) Merge pull request from EpicCash/fix/floonet-foundation
  - Fix(floonet): Make `foundation_floonet.json` the same as `foundation.json`
  - [#34](https://github.com/EpicCash/epic/issues/34) Merge pull request from EpicCash/feature/create-block-fees-submodule
  - Move the BlockFeeds type a separate location for better reusage: the wallets can use it there
  - Increasing the FLOONET_FIRST_HARD_FORK to 25800
  - [#33](https://github.com/EpicCash/epic/issues/33) Merge pull request from EpicCash/fix/cmake-error
  - [#32](https://github.com/EpicCash/epic/issues/32) Merge pull request from EricShimizuKarbstein/update-docs
  - Update randomx-rust
  - Update build instructions
  - Fix: more chunk size for header sync
  - Fix: add missing shutdown info in cmd mode
  - Refactor: update zeroise macro keychain, utils
  - [#29](https://github.com/EpicCash/epic/issues/29) Update epic cargo docs and wrong comments in code
  - [#28](https://github.com/EpicCash/epic/issues/28) Fix: update some deps in keychain
  - [#27](https://github.com/EpicCash/epic/issues/27) Merge pull request from hdmark/patch-1
  - Fix: update header sync; remove dublicate check bad header; remove debug output from graph weight
  - Fix: move reset stalling, sync peer reset
  - remove unused debug output
  - change header sync ban action
  - Fix: windows foundation height
  - [#26](https://github.com/EpicCash/epic/issues/26) Removing verifier caches
  - make small peers sync improvements
  - Fix: make header sync faster
  - update foundation sha for unix
  - update foundation hash; peers v2 to v3 sync
  - P2P: better handling of mpsc channel and initial handshake messages with other peers
  - Chain: Better output PMMR handling and Chain DB index. Improve stability and DB crashes.
  - New Header Version 7 and better Header Version Switch
  - Tui:
    - Some TUI improvements.
    - Lists can be scrolled now.
  - Split of File and Console  output and cleanup
  - API:
    - API enhancements and new API v3 for Wallet communication


# epic (3.0.0-beta)

  - changing algo mix for next block era 3
  - split stdout logformat from file log; stdout cosmetics


# epic (3.0.0-alpha-6)

  - rewind #3045
  - rewind Read header_head and sync_head from header MMRs directly
  - addd missing chain setup
  - additional code changes v3
  - pipe changes
  - add header hash to header info
  - check store for block sums
  - remove fixed size
  - change db store consts
  - add missing code v3
  - change zeroise to 1.3.x
  - change zerois to 1.x
  - change floonet era height; fix MAINNET_DIFFICULTY_ERA height
  - change current header version to 6
  - change cursive version
  - alpha6 finalize for testing
  - change stratum rpc id value
  - change floonet seed
  - fix v3 #3268


# epic (3.0.0-alpha-5)

  - change floonet first hardfork to header version 7 to blockheight 5760
  - change header back to v7
  - fix windows compatibility
  - add badblock script from 2.15;cursive tui update;version checker; set to v3.0.0
  - update ser because of serde::export private
  - Merge pull request #15 from johanneshahn/develop
  - Update build.md
  - Update build.md
  - Add missing 222 to FOUNDATION_LEVY
  - add missing foundation levy 222
  - API client basic authorization.
  - Adding cucumber tests - trying to spend foundation coins
  - Merge pull request #7 from EpicCash/fix-foudation-levy


# epic (3.0.0-alpha-4)

  - Merge branch 'fix-emission-schedule' into 'develop'
  - Update emission schedule
  - Merge branch 'fix--duration-tui' into 'develop'
  - Fix wrong duration on tui
  - Merge branch 'fix-floonet-foundation' into 'develop'
  - Fix floonet foundation on windows


# epic (3.0.0-alpha-3)

  - Fixing the communication between server and wallet
  - Update build.md -> remove "brew install llvm" for build with Mac. Mac ships own llvm


# epic (3.0.0-alpha-2)

  - Fixing conflict with EpicWallet


# epic (3.0.0-alpha-1)

  - This patch only affects floonet
  - Integrates grin 3.0.0 codebase into ours
  - Fixes some synchronization problems
  - Adjusts the difficulty adjustment a bit
  - Introduces a hard-fork in floonet at height 2880

# epic (3.0.0-alpha)

- Merge branch 'develop'
- Merge branch '1st-hard-fork' into develop
- Add foundation for floonet
- Add height for mainnet hardfork

config/src/config.rs
// Copyright 2018 The Grin Developers
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Configuration file management

use dirs;
use rand::distributions::{Alphanumeric, Distribution};
use rand::thread_rng;
use std::env;
use std::fs::{self, File};
use std::io::prelude::*;
use std::io::BufReader;
use std::io::Read;
use std::path::PathBuf;
use toml;

use crate::comments::insert_comments;
use crate::core::global;
use crate::p2p;
use crate::servers::ServerConfig;
use crate::types::{ConfigError, ConfigMembers, GlobalConfig};
use crate::util::LoggingConfig;

/// The default file name to use when trying to derive
/// the node config file location
pub const SERVER_CONFIG_FILE_NAME: &'static str = "epic-server.toml";
const SERVER_LOG_FILE_NAME: &'static str = "epic-server.log";
const EPIC_HOME: &'static str = ".epic";
const EPIC_CHAIN_DIR: &'static str = "chain_data";
/// Node API secret
pub const API_SECRET_FILE_NAME: &'static str = ".api_secret";

fn get_epic_path(chain_type: &global::ChainTypes) -> Result<PathBuf, ConfigError> {
	// Check if epic dir exists
	let mut epic_path = match dirs::home_dir() {
		Some(p) => p,
		None => PathBuf::new(),
	};
	epic_path.push(EPIC_HOME);
	epic_path.push(chain_type.shortname());
	// Create if the default path doesn't exist
	if !epic_path.exists() {
		fs::create_dir_all(epic_path.clone())?;
	}
	Ok(epic_path)
}

fn check_config_current_dir(path: &str) -> Option<PathBuf> {
	let p = env::current_dir();
	let mut c = p.ok()?;
	c.push(path);
	if c.exists() {
		return Some(c);
	}
	None
}

/// Create file with api secret
pub fn init_api_secret(api_secret_path: &PathBuf) -> Result<(), ConfigError> {
	let mut api_secret_file = File::create(api_secret_path)?;
	let api_secret: String = Alphanumeric
		.sample_iter(&mut thread_rng())
		.take(20)
		.collect();
	api_secret_file.write_all(api_secret.as_bytes())?;
	Ok(())
}

/// Check if file contains a secret and nothing else
pub fn check_api_secret(api_secret_path: &PathBuf) -> Result<(), ConfigError> {
	let api_secret_file = File::open(api_secret_path)?;
	let buf_reader = BufReader::new(api_secret_file);
	let mut lines_iter = buf_reader.lines();
	let first_line = lines_iter.next();
	if first_line.is_none() || first_line.unwrap().is_err() {
		fs::remove_file(api_secret_path)?;
		init_api_secret(api_secret_path)?;
	}
	Ok(())
}

/// Check that the api secret file exists and is valid
fn check_api_secret_file(chain_type: &global::ChainTypes) -> Result<(), ConfigError> {
	let epic_path = get_epic_path(chain_type)?;
	let mut api_secret_path = epic_path.clone();
	api_secret_path.push(API_SECRET_FILE_NAME);
	if !api_secret_path.exists() {
		init_api_secret(&api_secret_path)
	} else {
		check_api_secret(&api_secret_path)
	}
}

/// Handles setup and detection of paths for node
pub fn initial_setup_server(chain_type: &global::ChainTypes) -> Result<GlobalConfig, ConfigError> {
	check_api_secret_file(chain_type)?;
	// Use config file if current directory if it exists, .epic home otherwise
	if let Some(p) = check_config_current_dir(SERVER_CONFIG_FILE_NAME) {
		GlobalConfig::new(p.to_str().unwrap())
	} else {
		// Check if epic dir exists
		let epic_path = get_epic_path(chain_type)?;

		// Get path to default config file
		let mut config_path = epic_path.clone();
		config_path.push(SERVER_CONFIG_FILE_NAME);

		// Spit it out if it doesn't exist
		if !config_path.exists() {
			let mut default_config = GlobalConfig::for_chain(chain_type);
			// update paths relative to current dir
			default_config.update_paths(&epic_path);
			default_config.write_to_file(config_path.to_str().unwrap())?;
		}

		GlobalConfig::new(config_path.to_str().unwrap())
	}
}

/// Returns the defaults, as strewn throughout the code
impl Default for ConfigMembers {
	fn default() -> ConfigMembers {
		ConfigMembers {
			server: ServerConfig::default(),
			logging: Some(LoggingConfig::default()),
		}
	}
}

impl Default for GlobalConfig {
	fn default() -> GlobalConfig {
		GlobalConfig {
			config_file_path: None,
			members: Some(ConfigMembers::default()),
		}
	}
}

impl GlobalConfig {
	/// Same as GlobalConfig::default() but further tweaks parameters to
	/// apply defaults for each chain type
	pub fn for_chain(chain_type: &global::ChainTypes) -> GlobalConfig {
		let mut defaults_conf = GlobalConfig::default();
		let mut defaults = &mut defaults_conf.members.as_mut().unwrap().server;
		defaults.chain_type = chain_type.clone();

		match *chain_type {
			global::ChainTypes::Mainnet => {}
			global::ChainTypes::Floonet => {
				defaults.api_http_addr = "127.0.0.1:13413".to_owned();
				defaults.p2p_config.port = 13414;
				defaults
					.stratum_mining_config
					.as_mut()
					.unwrap()
					.stratum_server_addr = Some("127.0.0.1:13416".to_owned());
				defaults
					.stratum_mining_config
					.as_mut()
					.unwrap()
					.wallet_listener_url = "http://127.0.0.1:13415".to_owned();
			}
			global::ChainTypes::UserTesting => {
				defaults.api_http_addr = "127.0.0.1:23413".to_owned();
				defaults.p2p_config.port = 23414;
				defaults.p2p_config.seeding_type = p2p::Seeding::None;
				defaults
					.stratum_mining_config
					.as_mut()
					.unwrap()
					.stratum_server_addr = Some("127.0.0.1:23416".to_owned());
				defaults
					.stratum_mining_config
					.as_mut()
					.unwrap()
					.wallet_listener_url = "http://127.0.0.1:23415".to_owned();
			}
			global::ChainTypes::AutomatedTesting => {
				panic!("Can't run automated testing directly");
			}
		}
		defaults_conf
	}

	/// Requires the path to a config file
	pub fn new(file_path: &str) -> Result<GlobalConfig, ConfigError> {
		let mut return_value = GlobalConfig::default();
		return_value.config_file_path = Some(PathBuf::from(&file_path));

		// Config file path is given but not valid
		let config_file = return_value.config_file_path.clone().unwrap();
		if !config_file.exists() {
			return Err(ConfigError::FileNotFoundError(String::from(
				config_file.to_str().unwrap(),
			)));
		}

		// Try to parse the config file if it exists, explode if it does exist but
		// something's wrong with it
		return_value.read_config()
	}

	/// Read config
	fn read_config(mut self) -> Result<GlobalConfig, ConfigError> {
		let mut file = File::open(self.config_file_path.as_mut().unwrap())?;
		let mut contents = String::new();
		file.read_to_string(&mut contents)?;
		let decoded: Result<ConfigMembers, toml::de::Error> = toml::from_str(&contents);
		match decoded {
			Ok(gc) => {
				self.members = Some(gc);
				return Ok(self);
			}
			Err(e) => {
				return Err(ConfigError::ParseError(
					String::from(
						self.config_file_path
							.as_mut()
							.unwrap()
							.to_str()
							.unwrap()
							.clone(),
					),
					String::from(format!("{}", e)),
				));
			}
		}
	}

	/// Update paths
	pub fn update_paths(&mut self, epic_home: &PathBuf) {
		// need to update server chain path
		let mut chain_path = epic_home.clone();
		chain_path.push(EPIC_CHAIN_DIR);
		self.members.as_mut().unwrap().server.db_root = chain_path.to_str().unwrap().to_owned();
		let mut secret_path = epic_home.clone();
		secret_path.push(API_SECRET_FILE_NAME);
		self.members.as_mut().unwrap().server.api_secret_path =
			Some(secret_path.to_str().unwrap().to_owned());
		// TODO-FOUNDATION
		// let mut foundation_path = epic_home.clone();
		// foundation_path.push(
		// 	self.members
		// 		.as_ref()
		// 		.unwrap()
		// 		.server
		// 		.foundation_path
		// 		.clone(),
		// );
		// foundation_path.push("foundation.json");
		let foundation_path = if cfg!(windows) {
			let mut exec_path = std::env::current_exe().expect("Filesystem failure");
			exec_path.pop();
			exec_path.push("foundation.json");
			exec_path.to_str().expect("Invalid unicode path").to_owned()
		} else {
			"/usr/share/epic/foundation.json".to_owned()
		};

		// self.members.as_mut().unwrap().server.foundation_path =
		// 	foundation_path.to_str().unwrap().to_owned();

		self.members.as_mut().unwrap().server.foundation_path = foundation_path;
		let mut log_path = epic_home.clone();
		log_path.push(SERVER_LOG_FILE_NAME);
		self.members
			.as_mut()
			.unwrap()
			.logging
			.as_mut()
			.unwrap()
			.log_file_path = log_path.to_str().unwrap().to_owned();
	}

	/// Enable mining
	pub fn stratum_enabled(&mut self) -> bool {
		return self
			.members
			.as_mut()
			.unwrap()
			.server
			.stratum_mining_config
			.as_mut()
			.unwrap()
			.enable_stratum_server
			.unwrap();
	}

	/// Serialize config
	pub fn ser_config(&mut self) -> Result<String, ConfigError> {
		let encoded: Result<String, toml::ser::Error> =
			toml::to_string(self.members.as_mut().unwrap());
		encoded.map_err(|e| ConfigError::SerializationError(String::from(format!("{}", e))))
	}

	/// Write configuration to a file
	pub fn write_to_file(&mut self, name: &str) -> Result<(), ConfigError> {
		let conf_out = self.ser_config()?;
		let conf_out = insert_comments(conf_out);
		let mut file = File::create(name)?;
		file.write_all(conf_out.as_bytes())?;
		Ok(())
	}
}
